### PR TITLE
Refactor rnn models for the new standards

### PR DESCRIPTION
**Pull request by ktangsali**
# PhysicsNeMo Pull Request

## Description
Refactors the RNN models to the new coding standards. 

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

**Comment by ktangsali** [comment]
> ### Additional Comments (3)
> 1. `physicsnemo/models/rnn/rnn_one2many.py`, line 100-103 ([link](/nvidia/physicsnemo/blob/e5adc09e73950fbc717d87f32471eb1e6f49500e/physicsnemo/models/rnn/rnn_one2many.py#L100-L103))
>    **syntax:** output shape mismatch in docstring example - the example shows input `torch.randn(4, 6, 1, 16, 16)` which is [N=4, C=6, T=1, H=16, W=16], but the output shows `torch.Size([4, 6, 16, 16, 16])` which is [N=4, C=6, T=16, H=16, W=16]. the spatial dimensions should remain [16, 16], not become [16, 16, 16]
> 2. `physicsnemo/nn/conv_layers.py`, line 258-263 ([link](/nvidia/physicsnemo/blob/e5adc09e73950fbc717d87f32471eb1e6f49500e/physicsnemo/nn/conv_layers.py#L258-L263))
>    **logic:** padding order is incorrect for 2D input - the current code applies padding as `[pad_h // 2, pad_h - pad_h // 2, pad_w // 2, pad_w - pad_w // 2]` but F.pad expects padding in reverse dimension order: `[left, right, top, bottom]` which should be `[pad_w // 2, pad_w - pad_w // 2, pad_h // 2, pad_h - pad_h // 2]`
> 3. `physicsnemo/nn/conv_layers.py`, line 432-437 ([link](/nvidia/physicsnemo/blob/e5adc09e73950fbc717d87f32471eb1e6f49500e/physicsnemo/nn/conv_layers.py#L432-L437))
>    **logic:** padding/cropping order is incorrect for 2D input - torch.nn.functional.pad expects padding in reverse dimension order. current code uses `[pad_h // 2 : ..., pad_w // 2 : ...]` but the dimensions should be reversed to match width then height
>    can you verify the expected padding order matches your framework's conventions for transpose convolutions?
> 
> 7 files reviewed, 3 comments
> 
> [Edit Code Review Agent Settings](https://app.greptile.com/review/github) | [Greptile](https://greptile.com?utm_source=greptile_expert&utm_medium=github&utm_campaign=code_reviews&utm_content=physicsnemo_1306)

Fixed point 2 and 3. For point 1, the output is correct. [4, 6, 16, 16, 16] means batch=4, channels=6, timesteps=16, height=16, width=16

**Comment by coreyjadams** [comment]
I think this looks good to me.

When UNet is updated (Which is currently assigned to @peterdsharpe ...) we might consider consolidating some of the "ConvBlock" and "ConvLayer" etc implementation.  But this one came first so no need to hold it up :)

File: test/models/rnn/test_rnn.py
Comment (CharlelieLrt) [inline test/models/rnn/test_rnn.py:1]: IMO there are a few things missing in the current tests:

- The tests for the constructor should check the values of the public attributes (e.g. for `model = One2ManyRNN(...)`, the values of `model.nr_tsteps`, `model.nr_residual_blocks`, etc...)

- There should be a test that loads a checkpoint file. Unlike existing tests using checkpoints, the checkpoint file *should be part of the test suite*. It **should not** be created at runtime, otherwise we are not testing the ability to load checkpoints generated with older version of the codebase, which is the principal source of failure for loading checkpoints.

File: physicsnemo/nn/conv_layers.py [hunk 1]
```diff
@@ -282,3 +310,671 @@ def forward(self, x):
...
+                self.kernel_size,
+                self.stride,
+                bias=True,
+            )
+        else:
+            raise ValueError("Only 1D, 2D and 3D dimensions are supported")
+
+        self.reset_parameters()
+
+    def exec_activation_fn(
```
Comment (CharlelieLrt) [inline physicsnemo/nn/conv_layers.py]: Is this method expected to be externally exposed or is it for internal use? In the latter case it should be semi-private. This applies to all semi-private methods and attributes in these layers (e.g. `reset_parameters` below, etc.)

File: test/models/rnn/test_rnn_layers.py
Comment (CharlelieLrt) [inline test/models/rnn/test_rnn_layers.py:1]: The tests here only check the shape of the outputs. Would it be possible to have some non-regression tests that also check the values with `validate_accuracy`, the same way as it is done for the models?

Comment (ktangsali) [reply test/models/rnn/test_rnn_layers.py:1]: Addressed in [e2cf412](https://github.com/NVIDIA/physicsnemo/pull/1306/commits/e2cf4122618520ef3284e27bfdd4efdedd0021e3)

File: physicsnemo/nn/conv_layers.py (hunk 1 above)
Comment (ktangsali) [reply physicsnemo/nn/conv_layers.py]: Addressed in [7541105](https://github.com/NVIDIA/physicsnemo/pull/1306/commits/7541105265b97122099a7908eec135a4e332d7d0)

File: test/models/rnn/test_rnn.py
Comment (ktangsali) [reply test/models/rnn/test_rnn.py:1]: Addressed in [7541105](https://github.com/NVIDIA/physicsnemo/pull/1306/commits/7541105265b97122099a7908eec135a4e332d7d0)

**Comment by coreyjadams** [comment]
> There should be a test that loads a checkpoint file. Unlike existing tests using checkpoints, the checkpoint file should be part of the test suite.

@CharlelieLrt I like this idea a lot, but I'm thinking about having [~15 models] x [ checkpoint or two each] x [several MB per checkpoint or more, per model] leading to repo bloat.  Do we want to keep these checkpoint files with `git lfs` or perhaps a separate repo we can use as a `git submodule` during testing?

**Comment by CharlelieLrt** [comment]
@coreyjadams agree the size of the combined .mdlus checkpoints could be a problem. When I did that for other tests I made sure to use very tiny checkpoints that were all around 10KB to 100KB. I'm not sure if that would more acceptable (or even possible for some models). 

> a separate repo we can use as a git submodule during testing?

Agree, that would be a good solution

**Review by CharlelieLrt — APPROVED** [review]
LGTM!

**Comment by ktangsali** [comment]
> @coreyjadams agree the size of the combined .mdlus checkpoints could be a problem. When I did that for other tests I made sure to use very tiny checkpoints that were all around 10KB to 100KB. I'm not sure if that would more acceptable (or even possible for some models).
> 
> > a separate repo we can use as a git submodule during testing?
> 
> Agree, that would be a good solution

I like the idea of a separate repo for assets too. Let's address creation of a different repo for assets in a separate PR though! For now I have modified the checkpoints to be in the order of 10s of KBs.

**Comment by coreyjadams** [comment]
@ktangsali @CharlelieLrt  sounds good - we'll aim for a test-data repo or similar we can use for testing (I know there is already a private one) but since this isn't really a "functionality" issue it might have to wait a bit.
